File: Algorithms/JavaScript/min-depth-binary-tree.js
/**
 * Given a binary tree, find its minimum depth.
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 *
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
  if (root === null) {
    return 0;
  } else if (root.left === null && root.right === null) {
    return 1;
  } else if (root.left === null) {
    return 1 + minDepth(root.right);
  } else if (root.right === null) {
    return 1 + minDepth(root.left);
  } else {
    return 1 + Math.min(minDepth(root.left), minDepth(root.right));
  }
};

File: Algorithms/Ruby/min-depth-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}

def min_depth_binary_tree(root)
  return 0 unless root
  return 1 unless root.left && root.right
  return 1 + min_depth_binary_tree(root.left) unless root.right
  return 1 + min_depth_binary_tree(root.right) unless root.left

  1 + [min_depth_binary_tree(root.left), min_depth_binary_tree(root.right)].min
end
